# Project Description

**I've been given a database of one of the services competing in books apps market. It contains data on books, publishers, authors, customer ratings and reviews of books. My task is to analyze them and to generate a value proposition for a new product.**

**In order to do that firstly I need to:**

**-Study the tables by printing them, check if there are missing values and duplicates.**

**-Answering the questions bellow by making SQL queries for each task, and outputting the results.**

**-Describing my conclusions for each of the tasks.**

# Describe the goals of the study

**Since the coronavirus took the entire world by surprise, and changed everyone's daily routine. More startups rushed to develop new apps for book lovers. My goals for this project is to find which books were recently released (after the year 2000), number of user reviews and the average rating for each book. 
I also need to identify the publisher that has released the greatest number of books. And to detect the author with the highest average book rating and more.**

In [1]:
# import libraries
import pandas as pd
import warnings
from sqlalchemy import create_engine
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
warnings.simplefilter(action='ignore', category = FutureWarning)
from sqlalchemy import create_engine
%pip install psycopg2-binary
pd.set_option('display.max_colwidth', None)

Note: you may need to restart the kernel to use updated packages.


In [2]:
db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
books = pd.read_sql('SELECT * FROM books', con=engine)
authors = pd.read_sql('SELECT * FROM authors', con=engine)
publishers = pd.read_sql('SELECT * FROM publishers', con=engine)
ratings = pd.read_sql('SELECT * FROM ratings', con=engine)
reviews = pd.read_sql('SELECT * FROM reviews', con=engine)

# Study the tables


# Books

In [4]:
query = """
SELECT *
FROM books
LIMIT 5;"""
pd.io.sql.read_sql(query, con=engine)

book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
books.info()
books.tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id  \
995      996        571   
996      997        454   
997      998        201   
998      999         94   
999     1000        509   

                                                                                title  \
995                                          Wyrd Sisters (Discworld  #6; Witches #2)   
996                                                       Xenocide (Ender's Saga  #3)   
997                                                                   Year of Wonders   
998                                                       You Suck (A Love Story  #2)   
999  Zen and the Art of Motorcycle Maintenance: An Inquiry Into Values (Phaedrus  #1)   

     num_pages publication_date  publisher_id  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143

In [6]:
books.isnull().sum()

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

In [7]:
books.duplicated().sum()

0

# Authors

In [8]:
query = """
SELECT *
FROM authors
LIMIT 5;"""
pd.io.sql.read_sql(query, con=engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
authors.info()
authors.tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id                                author
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

In [10]:
authors.isnull().sum()

author_id    0
author       0
dtype: int64

In [11]:
authors.duplicated().sum()

0

# Publishers

In [12]:
query = """
SELECT *
FROM publishers
LIMIT 5;"""
pd.io.sql.read_sql(query, con=engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [13]:
publishers.info()
publishers.tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id                   publisher
335           336  Workman Publishing Company
336           337                  Wyatt Book
337           338       Yale University Press
338           339                    Yearling
339           340              Yearling Books

In [14]:
publishers.isnull().sum()

publisher_id    0
publisher       0
dtype: int64

In [15]:
publishers.duplicated().sum()

0

# Ratings

In [16]:
query = """
SELECT *
FROM ratings
LIMIT 5;"""
pd.io.sql.read_sql(query, con=engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [17]:
ratings.info()
ratings.tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id        username  rating
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

In [18]:
ratings.isnull().sum()

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

In [19]:
ratings.duplicated().sum()

0

# Reviews

In [20]:
query = """
SELECT *
FROM reviews
LIMIT 5;"""
pd.io.sql.read_sql(query, con=engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

In [21]:
reviews.info()
reviews.tail()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id  book_id        username  \
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                                                                                                           text  
2788                                                          Later hospital turn easy community. Fact same enter true practice student system.  
2789                                                                                           Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy population. Pm vote take. Long born close thought person old. Visit ever school manage about.  
2791                                           Or western offer wonder ask. More hear phone foreign amount within cell. Ball poor hot that kid.  
2792                                            Republican staff bit eat material measure plan. Pretty issue claim rate push energy couple bit.

In [22]:
reviews.isnull().sum()

review_id    0
book_id      0
username     0
text         0
dtype: int64

In [23]:
reviews.duplicated().sum()

0

**The datasets seem great, no missing values nor duplicates. Now it's time to answer the questions.**

*Find the number of books released after January 1, 2000.*

In [24]:
books_request= """
SELECT 
      publication_date
FROM   books
WHERE  
      publication_date > '2000-01-01';"""

In [25]:
pd.io.sql.read_sql(books_request, con = engine)

publication_date
0         2005-11-01
1         2003-05-22
2         2010-12-21
3         2006-10-10
4         2006-07-04
..               ...
814       2006-09-12
815       2001-02-06
816       2002-04-30
817       2007-01-16
818       2006-04-25

[819 rows x 1 columns]

**We have 819 books released after January 1, 2000.**

*Find the number of user reviews and the average rating for each book*

In [26]:
reviews_request = """SELECT 
  books.book_id,
  books.title,
  AVG(rating) AS average_rating,
  COUNT(DISTINCT review_id) AS number_reviews 
FROM books
LEFT JOIN reviews USING(book_id)
LEFT JOIN ratings USING(book_id)
GROUP BY books.book_id
ORDER BY average_rating DESC, number_reviews DESC, title;
"""

In [27]:
pd.io.sql.read_sql(reviews_request, con = engine)

book_id                                                           title  \
0         17                                   A Dirty Job (Grim Reaper  #1)   
1        444                    Moneyball: The Art of Winning an Unfair Game   
2        553                         School's Out—Forever (Maximum Ride  #2)   
3         20                           A Fistful of Charms (The Hollows  #4)   
4         55                       A Woman of Substance (Emma Harte Saga #1)   
..       ...                                                             ...   
995      915  The World Is Flat: A Brief History of the Twenty-first Century   
996      202                                                   Drowning Ruth   
997      316                               His Excellency: George Washington   
998      371                                                           Junky   
999      303                                            Harvesting the Heart   

     average_rating  number_reviews  
0              5.00               4  
1              5.00               3  
2              5.00               3  
3              5.00               2  
4              5.00               2  
..              ...             ...  
995            2.25               3  
996            2.00               3  
997            2.00               2  
998            2.00               2  
999            1.50               2  

[1000 rows x 4 columns]

In [28]:
average_ratings = """
SELECT 
  books.book_id, 
  books.title,
  AVG(rating) AS average_rating,
  COUNT(DISTINCT(rating_id)) AS number_ratings
FROM books
LEFT JOIN ratings USING(book_id)
GROUP BY books.book_id
ORDER BY average_rating DESC, number_ratings DESC;"""

In [29]:
pd.io.sql.read_sql(average_ratings, con = engine)

book_id                                                           title  \
0         17                                   A Dirty Job (Grim Reaper  #1)   
1        553                         School's Out—Forever (Maximum Ride  #2)   
2        444                    Moneyball: The Art of Winning an Unfair Game   
3        347            In the Hand of the Goddess (Song of the Lioness  #2)   
4        993                                                           Women   
..       ...                                                             ...   
995      915  The World Is Flat: A Brief History of the Twenty-first Century   
996      202                                                   Drowning Ruth   
997      371                                                           Junky   
998      316                               His Excellency: George Washington   
999      303                                            Harvesting the Heart   

     average_rating  number_ratings  
0              5.00               4  
1              5.00               4  
2              5.00               3  
3              5.00               3  
4              5.00               2  
..              ...             ...  
995            2.25               4  
996            2.00               3  
997            2.00               2  
998            2.00               2  
999            1.50               2  

[1000 rows x 4 columns]

**First plot we can see that 'A Dirty Job (Grim Reaper #1)' had the highest number of reviews (4) with the highest average of ratings(5) (there were other books with the same average but not with the same number of reviews like 'Moneyball: The Art of Winning an Unfair Game' who got only 3 reviews.**

**Second plot 'A Dirty Job (Grim Reaper #1)' still on top with 4 ratings. Plus 'School's Out—Forever (Maximum Ride #2)' had the same results(while in the first plot he was number 3 with 3 number reviews.**

*Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).*

In [30]:
greatest_publishers = """
SELECT 
      books.publisher_id,
      publishers.publisher,
      COUNT(books.title) as books_number_count
FROM publishers
INNER JOIN books ON publishers.publisher_id = books.publisher_id
WHERE num_pages > 50
GROUP BY books.publisher_id, publishers.publisher
ORDER BY books_number_count DESC
LIMIT 1;"""

In [31]:
pd.io.sql.read_sql(greatest_publishers, con = engine)

publisher_id      publisher  books_number_count
0           212  Penguin Books                  42

**From this plot we can tell that the greatest publisher is Penguin Books with with ID of 212, with book count of 42.** 

*Identify the author with the highest average book rating (look only at books with at least 50 ratings).*

In [32]:
author_highest = """
SELECT 
  authors.author,
  AVG(ratings.rating) AS avg_rating
FROM books 
INNER JOIN authors ON books.author_id = authors.author_id
INNER JOIN (
  SELECT 
    book_id,
    AVG(ratings.rating) AS book_avg_rating
  FROM ratings
  GROUP BY book_id
  HAVING COUNT(ratings.rating_id) >= 50
) AS book_ratings ON books.book_id = book_ratings.book_id
INNER JOIN ratings ON books.book_id = ratings.book_id
GROUP BY authors.author
ORDER BY avg_rating DESC;
"""

In [33]:
pd.io.sql.read_sql(author_highest, con = engine)

author  avg_rating
0                           J.K. Rowling/Mary GrandPré    4.287097
1                    Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                       J.R.R. Tolkien    4.246914
3                                    Louisa May Alcott    4.192308
4                                         Rick Riordan    4.080645
5                                      William Golding    3.901408
6                                        J.D. Salinger    3.825581
7             Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. Mowat    3.787879
9                                           Lois Lowry    3.750000
10                                           Dan Brown    3.741259
11             George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                     Stephenie Meyer    3.662500
13                                      John Steinbeck    3.622951

**The author with the highest average rating is J.K. Rowling/Mary GrandPré with average rating of 4.41 and rating count of 82. She had the highest average rating on her book 'Harry Potter and the Prisoner of Azkaban (Harry Potter #3)'.**

*Find the average number of text reviews among users who rated more than 50 books.*

In [34]:
average_number_text = """
SELECT AVG(sub.count_reviews) AS average_number_text
FROM (
  SELECT COUNT(DISTINCT(reviews.review_id)) AS count_reviews
  FROM reviews
  INNER JOIN ratings ON reviews.username = ratings.username
  GROUP BY ratings.username
  HAVING COUNT(DISTINCT(ratings.rating_id)) > 50
) sub;
"""

In [35]:
result = pd.io.sql.read_sql(average_number_text, con = engine)
result.round(2)

average_number_text
0                24.33

**The average number of text reviews among users who rated more than 50 books is 24.33 per user.**

# Describe your conclusions for each of the tasks

**1- About 819 books released after January 1, 2000.**

**2- We had 2 plots, first plot 'A Dirty Job (Grim Reaper #1)' had the highest number of reviews (4) with the highest average of ratings(5). Second plot 'A Dirty Job (Grim Reaper #1)' was still on top with 4 ratings. Plus 'School's Out—Forever (Maximum Ride #2)' had the same results(while in the first plot he was number 3 with 3 number reviews.**

**3- The greatest publisher is Penguin Books with with ID of 212, with book count of 42.** 

**4- The author with the highest average rating is J.K. Rowling/Mary GrandPré with average rating of 4.41 and rating count of         82. She had the highest average rating on her book 'Harry Potter and the Prisoner of Azkaban (Harry Potter #3)'.**

**5- The average number of text reviews among users who rated more than 50 books is 24.33 per user.**